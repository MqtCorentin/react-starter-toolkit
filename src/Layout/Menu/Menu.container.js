import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { useLocation, matchPath } from 'react-router-dom';
import { ToggleButton, Action } from '@axa-fr/react-toolkit-all';
import { Menu, menuItemPropType } from './Menu';
import MENU_ITEMS, { CLASS_BODY_MENU_OPEN } from './constants';

export const setPositionInit = ({ menuItems, pathname, matchPathFn = matchPath }) =>
  menuItems.map((navItem, index) => (pathname === navItem.url ? index : 0)).reduce((accumulator, currentValue) => accumulator + currentValue);

export const setToggleMenu = ({ setIsMenuVisible, isVisible, documentObj = document }) => {
  const { body } = documentObj;
  body.classList.toggle(CLASS_BODY_MENU_OPEN);
  setIsMenuVisible(!isVisible);
};

export const useMenuVisible = (initState = false) => {
  const [isVisible, setIsMenuVisible] = useState(initState);
  return { isVisible, setIsMenuVisible };
};

const MenuEnhanced = ({ menuItems, setPositionInitFn, setToggleMenuFn, useLocationFn, useMenuVisibleFn, ...rest }) => {
  const { pathname } = useLocationFn();
  const initPosition = setPositionInitFn({ menuItems, pathname });
  const { isVisible, setIsMenuVisible } = useMenuVisibleFn();

  const toggleMenu = useCallback(() => setToggleMenuFn({ setIsMenuVisible, isVisible }), [isVisible, setIsMenuVisible, setToggleMenuFn]);

  return (
    <>
      <Menu {...rest} menuItems={menuItems} isVisible={isVisible} handleClick={toggleMenu} initPosition={initPosition} />
      <ToggleButton idControl="mainmenu">
        <Action
          className="btn af-title-bar__mobile-menu af-btn--circle"
          id="togglemenu"
          icon="menu-hamburger"
          title="Toggle menu"
          onClick={toggleMenu}
        />
      </ToggleButton>
    </>
  );
};

export const menuEnhancedPropTypes = {
  menuItems: menuItemPropType,
  setPositionInitFn: PropTypes.func,
  setToggleMenuFn: PropTypes.func,
  useLocationFn: PropTypes.func,
  useMenuVisibleFn: PropTypes.func,
};

export const menuEnhancedDefaultProps = {
  menuItems: MENU_ITEMS,
  setPositionInitFn: setPositionInit,
  setToggleMenuFn: setToggleMenu,
  useLocationFn: useLocation,
  useMenuVisibleFn: useMenuVisible,
};

MenuEnhanced.propTypes = menuEnhancedPropTypes;
MenuEnhanced.defaultProps = menuEnhancedDefaultProps;

export default MenuEnhanced;
