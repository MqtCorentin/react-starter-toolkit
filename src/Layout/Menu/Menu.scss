@import '@axa-fr/react-toolkit-core/dist/assets/scss/core.scss';

.af-title-bar__mobile-menu {
  display: none;
  background-color: $color-axa;
}
@include media-breakpoint-down(sm) {
  .af-title-bar__mobile-menu {
    position: fixed;
    display: block;
    top: 84px;
    right: 0;
    z-index: 12;
    width: 2rem;
    height: 2rem;
  }
}
