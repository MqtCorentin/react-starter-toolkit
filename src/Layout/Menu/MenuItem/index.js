export { default, createMenu } from './MenuItem.container';
export { MenuItem, MenuItemPropTypes } from './MenuItem';
