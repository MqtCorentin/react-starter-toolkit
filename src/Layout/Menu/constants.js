import ROUTE_URL from 'Layout/constants';

export const CLASS_BODY_MENU_OPEN = 'af-menu-open';

const MENU_ITEMS = [
  {
    label: 'Accueil',
    url: ROUTE_URL.DASHBOARD,
  },
  {
    label: 'Membres',
    url: ROUTE_URL.MEMBERS,
  },
  {
    label: 'Forms',
    url: ROUTE_URL.FORM,
    children: [
      {
        label: 'Champs classiques',
        url: `${ROUTE_URL.FORM}${ROUTE_URL.FIELDS_CLASSICS}`,
      },
      {
        label: 'Champ datepicker',
        url: `${ROUTE_URL.FORM}${ROUTE_URL.DATEPICKER}`,
      },
      {
        label: 'Champ slider',
        url: `${ROUTE_URL.FORM}${ROUTE_URL.SLIDER}`,
      },
    ],
  },
];

export default MENU_ITEMS;
