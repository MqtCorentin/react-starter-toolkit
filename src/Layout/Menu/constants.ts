import ROUTE_URL from 'App/Routes/constants';

export const CLASS_BODY_MENU_OPEN = 'af-menu-open';

const MENU_ITEMS = [
  {
    label: 'Accueil',
    url: ROUTE_URL.SLASH,
  },
  {
    label: 'Membres',
    url: ROUTE_URL.MEMBERS,
  },
  {
    label: 'Rechercher',
    url: ROUTE_URL.SEARCHMEMBERS,
  },
  {
    label: 'Demos',
    children: [
      {
        label: 'Modal',
        url: ROUTE_URL.MODAL,
      },
      {
        label: 'Button',
        url: ROUTE_URL.BUTTON,
      },
      {
        label: 'Notification',
        url: ROUTE_URL.NOTIFICATION,
      },
    ],
  },
  {
    label: 'Pages',
    children: [
      {
        label: 'Empty page',
        url: ROUTE_URL.DASHBOARD,
      },
      {
        label: 'Not found',
        url: ROUTE_URL.NOTFOUND,
      },
      {
        label: 'Forbidden',
        url: ROUTE_URL.UNAUTHORIZE,
      },
    ],
  },
];

export default MENU_ITEMS;
