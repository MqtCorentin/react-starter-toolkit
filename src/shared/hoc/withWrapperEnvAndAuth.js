/**
 * This HOC is useful for testing components that need to be wwrapped by environment and authentication providers
 */
import React from 'react';
import { AuthenticationProvider } from '@axa-fr/react-oidc-context';
import { EnvironmentProvider } from 'App/Environment';

const oidc = {
  isEnabled: true,
  client_id: '0123456',
  redirect_uri: 'http://localhost:3000/authentication/callback',
  response_type: 'code',
  post_logout_redirect_uri: 'http://localhost:3000/logout',
  scope: 'openid profile email',
  authority: 'https://autority/',
  silent_redirect_uri: 'http://localhost:3000/authentication/silent_callback',
  automaticSilentRenew: true,
  loadUserInfo: true,
  triggerAuthFlow: true,
};

export default Component => props => (
  <EnvironmentProvider value={{ environment: { oidc } }}>
    <AuthenticationProvider configuration={oidc} isEnabled>
      <Component {...props} />
    </AuthenticationProvider>
  </EnvironmentProvider>
);
