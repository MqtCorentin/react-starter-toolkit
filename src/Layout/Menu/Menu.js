import React from 'react';
import PropTypes from 'prop-types';
import { NavBar } from '@axa-fr/react-toolkit-all';
import { MenuItemPropTypes, createMenu } from './MenuItem';
import './Menu.scss';

export const Menu = ({ menuItems, initPosition, isVisible, handleClick }) => (
  <NavBar positionInit={initPosition} isVisible={isVisible} onClick={handleClick}>
    {createMenu(menuItems)}
  </NavBar>
);

export const menuItemPropType = PropTypes.arrayOf(
  PropTypes.shape({
    ...MenuItemPropTypes,
  }),
);

export const MenuPropTypes = {
  menuItems: menuItemPropType.isRequired,
  initPosition: PropTypes.number,
};

export const MenuDefaultProps = {
  initPosition: 0,
};

Menu.propTypes = MenuPropTypes;
Menu.defaultProps = MenuDefaultProps;
