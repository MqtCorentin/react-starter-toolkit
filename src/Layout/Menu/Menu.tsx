import { NavBar } from '@axa-fr/react-toolkit-all';
import { createMenu, TMenuItemBase } from './MenuItem';
import './Menu.scss';

export type TMenu = {
  menuItems?: TMenuItemBase[];
  createMenuFn?: typeof createMenu;
  isVisible?: boolean;
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
  positionInit?: number;
};

export const Menu = ({ menuItems, createMenuFn = createMenu, ...navBarProps }: TMenu) => (
  <NavBar {...navBarProps}>{menuItems && createMenuFn(menuItems)}</NavBar>
);
