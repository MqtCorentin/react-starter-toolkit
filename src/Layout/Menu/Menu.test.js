import React from 'react';
import { render, act } from '@testing-library/react';
import { WrapperStaticRouter } from 'shared/testsUtils';
import withWrapperEnvAndAuth from 'shared/hoc/withWrapperEnvAndAuth';
import { Menu } from './Menu';
import MENU_ITEMS from './constants';

const MenuWithEnvAndAuth = withWrapperEnvAndAuth(Menu);

describe('<Menu/>', () => {
  it('Render <Menu/>', async () => {
    await act(async () => {
      const { asFragment } = render(<MenuWithEnvAndAuth menuItems={MENU_ITEMS} isVisible />, { wrapper: WrapperStaticRouter });
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
