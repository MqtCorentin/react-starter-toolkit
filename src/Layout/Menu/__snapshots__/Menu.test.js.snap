// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu/> Render <Menu/> 1`] = `
<DocumentFragment>
  <div
    class="af-nav-container"
  >
    <button
      aria-controls="mainmenu"
      aria-haspopup="true"
      class="mask fade show"
      id="togglemenu"
      type="button"
    />
    <nav
      aria-label="Menu principal"
      class="af-nav af-drawer left show"
      role="navigation"
    >
      <div
        class="container af-nav__wrapper"
      >
        <div
          class="af-nav__aside-header"
        >
          <h4
            class="af-nav__aside-title"
          >
            Menu
          </h4>
          <button
            aria-controls="mainmenu"
            aria-haspopup="true"
            class="af-nav__aside-close"
            id="togglemenu"
            type="button"
          >
            Ã—
          </button>
        </div>
        <ul
          aria-labelledby="togglemenu"
          class="af-nav__list "
          id="mainmenu"
          role="menubar"
        >
          <li
            class="af-nav__item af-nav__item--active "
            role="none"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Accueil"
              class="af-nav__link"
              href="/"
              role="menuitem"
              tabindex="0"
            >
              Accueil
            </a>
          </li>
          <li
            class="af-nav__item "
            role="none"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Membres"
              class="af-nav__link"
              href="/members"
              role="menuitem"
              tabindex="-1"
            >
              Membres
            </a>
          </li>
          <li
            class="af-nav__item af-nav__item--haschild "
            role="none"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Forms"
              class="af-nav__link"
              href="/form"
              role="menuitem"
              tabindex="-1"
            >
              Forms
            </a>
            <i
              class="glyphicon glyphicon-arrow-xs-down"
            />
            <ul
              aria-label=""
              class="af-nav__list af-nav__list--children"
              role="menu"
            >
              <li
                class="af-nav__item "
                role="none"
              >
                <a
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-label="Champs classiques"
                  class="af-nav__link"
                  href="/form/classics-fields"
                  role="menuitem"
                  tabindex="-1"
                >
                  Champs classiques
                </a>
              </li>
              <li
                class="af-nav__item "
                role="none"
              >
                <a
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-label="Champ datepicker"
                  class="af-nav__link"
                  href="/form/datepicker"
                  role="menuitem"
                  tabindex="-1"
                >
                  Champ datepicker
                </a>
              </li>
              <li
                class="af-nav__item "
                role="none"
              >
                <a
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-label="Champ slider"
                  class="af-nav__link"
                  href="/form/slider"
                  role="menuitem"
                  tabindex="-1"
                >
                  Champ slider
                </a>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </nav>
  </div>
</DocumentFragment>
`;
